packages/compiler/core/projector.ts

```diff
@@ -589,19 +594,49 @@ export function createProjector(
       if (projectionsByName.length === 0) continue;
       const targetNode =
         projectionApplication.direction === "from"
-          ? projectionsByName[0].from!
-          : projectionsByName[0].to!;
-      const projected = checker.project(projectedType, targetNode, projectionApplication.arguments);
-      if (projected !== projectedType) {
-        // override the projected type cache with the returned type
-        projectedTypes.set(baseType, projected);
-        return projected;
+          ? projectionsByName[0].from
+          : projectionsByName[0].to;
+
+      if (targetNode) {
+        const projected = checker.project(
+          projectedType,
+          targetNode,
+          projectionApplication.arguments
+        );
+        if (projected !== projectedType) {
+          // override the projected type cache with the returned type
+          projectedTypes.set(baseType, projected);
+          return projected;
+        }
       }
     }
 
     return projectedType;
   }
 
+  function applyPreProjection(type: Type): Type {
+    const inScopeProjections = getInScopeProjections();
+    for (const projectionApplication of inScopeProjections) {
+      const projectionsByName = type.projectionsByName(projectionApplication.projectionName);
+      if (projectionsByName.length === 0) continue;
+      const targetNode =
+        projectionApplication.direction === "from"
+          ? projectionsByName[0].preFrom
+          : projectionsByName[0].preTo;
+
+      if (targetNode) {
+        const projected = checker.project(type, targetNode, projectionApplication.arguments);
+        if (projected !== type) {
+          // override the projected type cache with the returned type
+          projectedTypes.set(type, projected);
```

Comment:
```suggestion
          projectedTypes.set(type, projected);
```

would this not cause any issues with referencing the previous type if you already made a reference to it?

Reply:
Yeah, you're right, this seems wrong. I wrote this so long ago and this got put aside until we resolved the design questions that I forget what the intent was here. I will look and add more tests.

Reply:
I don't think there was any intent, it looks like I left it mistakenly by copying from applyProjection to applyPreProjection.

Reply:
I have opened a can of worms testing this. Will file an issue. There are pre-existing problems, I think with return `<non-intrinsic>`.